File: frontend/lib/api/feed.js
import { apiFetch } from "./client";

/**
 * Paginated, filterable discovery feed — used by the /feed page.
 * Returns { items: [...], total: number }.
 */
export function listFeed({ limit = 12, offset = 0, signalType = null, sort = "recent", status = null } = {}) {
  const params = new URLSearchParams();
  params.set("limit", limit);
  params.set("offset", offset);
  if (signalType) params.set("signal_type", signalType);
  if (sort) params.set("sort", sort);
  if (status) params.set("status", status);
  return apiFetch(`/feed/signals?${params.toString()}`);
}

/** Single signal, for the /feed/[signalId] detail page. */
export function getFeedItem(signalId) {
  return apiFetch(`/feed/signals/${signalId}`);
}

/** Top N signals by confidence — used by the dashboard's Discovery Feed cards. */
export async function fetchSignals(limit = 4) {
  const { items } = await listFeed({ limit, sort: "confidence" });
  return items;
}

/** The 4 top stat cards on the dashboard. */
export function fetchStats() {
  return apiFetch(`/feed/stats`);
}

/** Runs the clustering pipeline on-demand — embeds every active case,
 * clusters them, and writes real research_signals rows. */
export function triggerScan() {
  return apiFetch(`/feed/scan`, { method: "POST" });
}

File: backend/app/routers/feed.py
"""
Feed router — powers the dashboard's Discovery Feed cards, the /feed list
page, and the /feed/[signalId] detail page.

GET  /feed/signals            -> paginated + filterable list ({items, total})
GET  /feed/signals/{signal_id} -> single signal
GET  /feed/stats               -> the 4 stat cards on the dashboard
POST /feed/scan                -> run the clustering pipeline on-demand,
                                   turning real cases into real signals
"""

from fastapi import APIRouter, Depends, Query

from app.deps.auth import get_current_user
from app.schemas.feed import FeedSignalsPage, FeedStats, ResearchSignalRead, ScanResult
from app.services import feed_service
from app.services.clustering_service import run_discovery_scan

router = APIRouter(prefix="/feed", tags=["feed"])


@router.get("/signals", response_model=FeedSignalsPage)
async def get_signals(
    limit: int = Query(12, ge=1, le=100),
    offset: int = Query(0, ge=0),
    signal_type: str | None = Query(None),
    sort: str = Query("recent"),  # 'recent' | 'confidence'
    status: str | None = Query("active"),
    user: dict = Depends(get_current_user),
):
    return feed_service.list_signals(
        limit=limit, offset=offset, signal_type=signal_type, sort=sort, signal_status=status
    )


@router.get("/signals/{signal_id}", response_model=ResearchSignalRead)
async def get_signal(signal_id: str, user: dict = Depends(get_current_user)):
    return feed_service.get_signal(signal_id)


@router.get("/stats", response_model=FeedStats)
async def get_stats(user: dict = Depends(get_current_user)):
    return feed_service.get_stats(hospital_id=user.get("hospital_id"))


@router.post("/scan", response_model=ScanResult)
async def scan(user: dict = Depends(get_current_user)):
    """Runs the clustering pipeline synchronously: embeds every active case,
    clusters them, and upserts research_signals + research_signal_cases.
    On a large case volume this should move to a background task — fine
    for on-demand use at current scale."""
    return await run_discovery_scan()